source/iterator/snapshot.go

```diff
@@ -0,0 +1,261 @@
+)
+
+// snapshotIterator - iterator which get snapshot data.
+// A "snapshot" is the state of a table data at a particular point in time when connector starts work.
+// The first time when the snapshot iterator starts work, it is gets max value from `orderingColumn` and saves
+// this value to position.
+// The snapshot iterator reads all rows, where `orderingColumn` values less or equal maxValue,
+// from the table in batches.
+// Values in the ordering column must be unique and suitable for sorting, otherwise, the snapshot won't work correctly.
+// Iterators saves last processed value from `orderingColumn` column to position to field `SnapshotLastProcessedVal`.
+// If snapshot stops it will parse position from last record and will
+// try gets row where `{{orderingColumn}} > {{position.SnapshotLastProcessedVal}}`.
+type snapshotIterator struct {
+	db   *sqlx.DB
+	rows *sqlx.Rows
+
+	// table - table name.
+	table string
+	// keys Names of columns what iterator use for setting key in record.
+	keys []string
+	// orderingColumn Name of column what iterator using for sorting data.
+	orderingColumn string
+	// maxValue max value from ordering column. Connector uses this variable like boundary value for snapshot.
+	maxValue any
+	// batchSize size of batch.
+	batchSize int
+	// position last recorded position.
+	position *position.Position
+	// columnTypes column types from table.
+	columnTypes map[string]string
+	// suffixName special suffix that connector uses for identify tracking table and triggers.
+	suffixName string
+}
+
+type snapshotParams struct {
+	db             *sqlx.DB
+	table          string
+	orderingColumn string
+	keys           []string
+	batchSize      int
+	position       *position.Position
+	columnTypes    map[string]string
+	suffixName     string
+}
+
+func newSnapshotIterator(
+	ctx context.Context,
+	snapshotParams snapshotParams,
+) (*snapshotIterator, error) {
+	var err error
+
+	it := &snapshotIterator{
+		db:             snapshotParams.db,
+		table:          snapshotParams.table,
+		keys:           snapshotParams.keys,
+		orderingColumn: snapshotParams.orderingColumn,
+		batchSize:      snapshotParams.batchSize,
+		position:       snapshotParams.position,
+		columnTypes:    snapshotParams.columnTypes,
+		suffixName:     snapshotParams.suffixName,
+	}
+
+	err = it.loadRows(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("load rows: %w", err)
+	}
+
+	if snapshotParams.position != nil {
+		it.maxValue = snapshotParams.position.SnapshotMaxValue
+	} else {
+		err = it.setMaxValue(ctx)
+		if err != nil {
+			return nil, fmt.Errorf("set max value: %w", err)
+		}
+	}
+
+	return it, nil
+}
+
+// HasNext check ability to get next record.
+func (i *snapshotIterator) HasNext(ctx context.Context) (bool, error) {
+	if i.rows != nil && i.rows.Next() {
+		return true, nil
+	}
+
+	if err := i.loadRows(ctx); err != nil {
+		return false, fmt.Errorf("load rows: %w", err)
```

Comment:
if we get error inside this function we get  "has next" in error message. Because it was wrapped
```
	switch {
	case c.snapshot != nil:
		hasNext, err := c.snapshot.HasNext(ctx)
		if err != nil {
			return false, fmt.Errorf("snapshot has next: %w", err)
		}
```
so we can something like it:
```
source has next: snapshot has next: load rows: some error
```